### ISSUE TITLE
[PROD] DP homepage sometimes loads pretty slowly (up to 5 seconds)

### ISSUE THREAD

**Issue by tihuan** [chanzuckerberg/corpora-data-portal#565]
Appetite: 2

What:
DP homepage sometimes takes up to 5 seconds to load all dataset. I haven't been able to reproduce the issue reliably, but at least both Bruce and I experience it a few times

Ask:
1. Set up an acceptance criterion for loading homepage under 1 second
2. Figure out the bottleneck and resolve the issue please 🙏

**Comment by maniarathi** [comment]
I suspect this might be due to the API fetching project details per project. In the future this landing page will go away so is it worthwhile spending the time to optimize it now. Or some basic caching would do the trick?

**Comment by tihuan** [comment]
Yeah the interesting thing is that it comes and goes, so I haven't been able to tell what the bottleneck is

Bruce was also saying we can possibly add Datadog to monitor the endpoints if that helps detect the slow down?

**Comment by tihuan** [comment]
Bruce's potential idea for API perf goal for /project and /project/{id}: Mean < 500ms, 90% < 1 second

**Comment by brianraymor** [comment]
@tihuan - why is a performance/tuning issue a bug?

**Comment by tihuan** [comment]
Oh I was just thinking that the perf issue is an anomaly instead of an expected behavior, so I lumped that into the bug category. But we could treat that as a tech debt or just a regular feature ticket?

**Comment by tihuan** [comment]
Looks like `/project` is not the problem (155ms), but `/project/{id}`! (At least for `dev` env API)

![Screen Shot 2020-08-31 at 3.41.09 PM.png](https://images.zenhubusercontent.com/5ebed8780d2d150fc281ad55/0f522c78-03aa-47f2-8769-aeaebcb7dbc6)

![Screen Shot 2020-08-31 at 3.38.48 PM.png](https://images.zenhubusercontent.com/5ebed8780d2d150fc281ad55/5fd354db-faba-40c3-b3f0-a1d857771798)

I saw Marcus mentioned something about keeping the lambda warm, but just wanted to provide more information in case it's helpful :D 

Thank you!

**Comment by tihuan** [comment]
@Bento007 and @MDunitz will pair on refining this ticket and help us explore a solution to keep the API calls under 1 second, and ensure the ticket has enough details for estimation 😄

Thanks both!

**Comment by MDunitz** [comment]
Trent and I located the api gateway latency metrices. The latency was extremely spiky indicating an issue with cold starts. Using locust we tested the dev api (see stats below[0]). We then turned on the api cache and retested[1] and found a 10x improvement. We recommend add caching through api gateway (this may be challenging due to how chalice and terraform interact). In terms of regression testing we recommend setting up cloudfront monitoring/alerting.

[0] no cache stats
 ```Name                                                          # reqs      # fails  |     Avg     Min     Max  Median  |   req/s failures/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /dp/v1/project                                               608     0(0.00%)  |     199      97    4469     130  |    8.05    0.00
 GET /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481          127     0(0.00%)  |     345     220    5067     270  |    1.68    0.00
 GET /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca          126     0(0.00%)  |     389     220    5070     260  |    1.67    0.00
 GET /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6          129     0(0.00%)  |     798     413    5463     500  |    1.71    0.00
 GET /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174          104     0(0.00%)  |     374     193    4534     230  |    1.38    0.00
 GET /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379          116     0(0.00%)  |     480     243    4985     280  |    1.53    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregated                                                      1210     0(0.00%)  |     340      97    5463     220  |   16.01    0.00
Response time percentiles (approximated)
 Type     Name                                                              50%    66%    75%    80%    90%    95%    98%    99%  99.9% 99.99%   100% # reqs
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 GET      /dp/v1/project                                                    130    140    140    150    200    320    590   4300   4500   4500   4500    608
 GET      /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481               270    280    290    300    510    660    840    850   5100   5100   5100    127
 GET      /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca               260    270    280    290    370    610   4500   4600   5100   5100   5100    126
 GET      /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6               500    520    530    540    720   4800   5200   5300   5500   5500   5500    129
 GET      /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174               230    240    250    250    320    660   4300   4500   4500   4500   4500    104
 GET      /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379               290    300    330    360    550    750   4800   4900   5000   5000   5000    116
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 None     Aggregated                                                        220    260    290    320    500    600   4300   4800   5300   5500   5500   1210
```


[1] With cache stats
``` Name                                                          # reqs      # fails  |     Avg     Min     Max  Median  |   req/s failures/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /dp/v1/project                                               568     0(0.00%)  |      63      36    1071      50  |   10.37    0.00
 GET /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481           98     0(0.00%)  |      66      43     558      52  |    1.79    0.00
 GET /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca           93     0(0.00%)  |      73      42     849      54  |    1.70    0.00
 GET /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6           91     0(0.00%)  |      68      43     570      53  |    1.66    0.00
 GET /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174           87     0(0.00%)  |      56      40     179      53  |    1.59    0.00
 GET /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379           93     0(0.00%)  |      70      45    1068      52  |    1.70    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregated                                                      1030     0(0.00%)  |      65      36    1071      51  |   18.81    0.00
Response time percentiles (approximated)
 Type     Name                                                              50%    66%    75%    80%    90%    95%    98%    99%  99.9% 99.99%   100% # reqs
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 GET      /dp/v1/project                                                     50     53     56     58     67     83    200    400   1100   1100   1100    568
 GET      /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481                52     55     59     60     71    150    430    560    560    560    560     98
 GET      /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca                54     57     60     62     73    150    540    850    850    850    850     93
 GET      /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6                53     58     61     63     72    180    200    570    570    570    570     91
 GET      /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174                53     56     59     60     65     75    130    180    180    180    180     87
 GET      /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379                52     55     59     59     71    130    210   1100   1100   1100   1100     93
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 None     Aggregated                                                         51     54     58     60     69    110    200    430   1100   1100   1100   1030```

**Comment by tihuan** [comment]
Love the comprehensive table and description!! 🔥 👍

**Comment by MDunitz** [comment]
That was all @Bento007 and his locust script!

**Comment by MDunitz** [comment]
Update 9/14 
Filed case with AWS support 

```When I turn on caching in the api gateway the wrong data is returned for various endpoints.

Without caching all of the endpoints return the correct data. When UI turn on caching it appears that the endpoints returns whatever data I request first.
For example if I turn on caching and then go to `https://api.dev.corpora.cziscience.com/dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481` I get the correct data set.
But if I try to go to https://api.dev.corpora.cziscience.com/dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174` it will return the `18e6337b-8c36-4977-9ac8-ca8aa5494481` dataset```

**Comment by MDunitz** [comment]
Notes from pairing with Trent
Turned on Provisioned Concurrency with 5 lambdas
Locust with xray, 30 users, 10 users spawn per second.
 Name                                                          # reqs      # fails  |     Avg     Min     Max  Median  |   req/s failures/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /dp/v1/project                                               433     0(0.00%)  |     273     120    4869     190  |    7.60    0.00
 GET /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481           96     0(0.00%)  |     491     274    4583     400  |    1.69    0.00
 GET /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca           80     0(0.00%)  |     739     295    5730     390  |    1.40    0.00
 GET /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6           93     0(0.00%)  |     873     510    2528     780  |    1.63    0.00
 GET /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174           92     0(0.00%)  |     457     244    4943     330  |    1.62    0.00
 GET /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379           78     0(0.00%)  |     491     274    1940     430  |    1.37    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregated                                                       872     0(0.00%)  |     442     120    5730     320  |   15.31    0.00
Response time percentiles (approximated)
 Type     Name                                                              50%    66%    75%    80%    90%    95%    98%    99%  99.9% 99.99%   100% # reqs
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 GET      /dp/v1/project                                                    190    210    230    250    420    600    770   1200   4900   4900   4900    433
 GET      /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481               400    450    480    500    710    790   1000   4600   4600   4600   4600     96
 GET      /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca               390    480    540    620    810   4900   5300   5700   5700   5700   5700     80
 GET      /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6               780    850    960   1000   1200   1400   2300   2500   2500   2500   2500     93
 GET      /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174               330    370    400    450    720    900   1300   4900   4900   4900   4900     92
 GET      /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379               430    470    510    550    750    830   1100   1900   1900   1900   1900     78
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 None     Aggregated                                                        320    400    490    570    770    940   1300   4500   5700   5700   5700    872

**Comment by tihuan** [comment]
Thanks so much for the constant updates for lurkers like me to follow along 😆 

The table above seems to show similar performance as https://github.com/chanzuckerberg/corpora-data-portal/issues/565#issuecomment-690734828 No cache stats[0]? Or maybe I'm not reading it correctly lol

**Comment by MDunitz** [comment]
Warmed cloudfront cache, no provisioned lambdas (through the cloudfront api `https://api.cellxgene.dev.single-cell.czi.technology/dp/v1/project/{uuid}`
```
 Name                                                          # reqs      # fails  |     Avg     Min     Max  Median  |   req/s failures/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /dp/v1/project                                               204     0(0.00%)  |     178     119     527     160  |    8.18    0.00
 GET /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481           36     0(0.00%)  |     329     268     420     310  |    1.44    0.00
 GET /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca           43     0(0.00%)  |     342     244     759     330  |    1.72    0.00
 GET /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6           40     0(0.00%)  |     694     555     977     650  |    1.60    0.00
 GET /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174           41     0(0.00%)  |     349     238    1359     280  |    1.64    0.00
 GET /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379           41     0(0.00%)  |     383     290     565     350  |    1.64    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregated                                                       405     0(0.00%)  |     298     119    1359     270  |   16.24    0.00
Response time percentiles (approximated)
 Type     Name                                                              50%    66%    75%    80%    90%    95%    98%    99%  99.9% 99.99%   100% # reqs
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 GET      /dp/v1/project                                                    160    180    180    190    260    310    360    400    530    530    530    204
 GET      /dp/v1/project/18e6337b-8c36-4977-9ac8-ca8aa5494481               320    360    370    380    390    420    420    420    420    420    420     36
 GET      /dp/v1/project/673637cf-dcb7-45e1-bb88-72a27c50c8ca               330    350    360    370    390    390    760    760    760    760    760     43
 GET      /dp/v1/project/8ad74eb4-c68f-4a4a-a17e-354f17605da6               700    750    750    760    820    900    980    980    980    980    980     40
 GET      /dp/v1/project/a55cac37-629f-4d28-9b3f-4ece7a7ac174               280    290    300    310    470    610   1400   1400   1400   1400   1400     41
 GET      /dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379               350    420    430    440    520    540    570    570    570    570    570     41
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 None     Aggregated                                                        270    310    350    380    570    730    770    870   1400   1400   1400    405

**Comment by MDunitz** [comment]
@Bento007 and I determined that using the new merged domain with cloudfront caching achieves the latency goals.

**Comment by MDunitz** [comment]
![image.png](https://images.zenhubusercontent.com/5b4523ca3f1e7333c932b3b1/c7143a82-1c23-40b5-8657-76f9dee6641f)

**Comment by Bento007** [comment]
Thank you @MDunitz for updating the ticket. 
## Solutions
### Stand up cloudfront in front of api-gateway and the lambda
#### Pros
- This results in api call under 200ms.
- We are already doing it

#### Cons
- must clear the cloudfront cache whenever a database entry is add or modified. This can be a slow process.
- need to verify that this will work as intended once authentication is in place.

### Enable Caching at on api-gateway
#### Pros
- results in api call under 500ms
- Able to control the cache time to live for each endpoint

#### Cons
- Will take more work to setup using chalice.
- The cached response could get stale when making databases.

### Reserve lambdas
#### Pro
- result in api calls under 500ms
- The database response is always up to date

#### Cons
- As load increases we will eventually run into the slow startup time of 5 seconds as more lambdas spin up.

### Optimize the spin-up time of Lambdas
#### Pro
- We could reduce the API spin up time to under 1 second.

#### Cons
- could require rebuilding of the backend server.

## Future Work
We must verify certain things still work as intended. Verify that cache control headers work as expected. Verify that Authentication works as expected.

## Conclusion
Any combination of the proposed solution will improve the load time of the frontend. Using cloudfront is the best solution for the time being. With that said we will need to verify cache control headers work, and authentication work moving forward.

**Comment by MDunitz** [comment]
Added latency tests (<500ms) to the cloudfront api for 
- https://api.cellxgene.dev.single-cell.czi.technology/dp/v1/project
- https://api.cellxgene.dev.single-cell.czi.technology/dp/v1/project/e7c47289-22fe-440e-ba24-a4434ac3f379



TODO 
- add tests to staging/prod env when those changes are merged

**Comment by tihuan** [comment]
Reopening this design ticket for next sprint to further our investigation 😄 Thank you!

**Comment by bmccandless** [comment]
@Bento007 has several options to fix, which would have different estimates. 

Another option (which could be used in combination with some of the others), is to implement a new API that gets all the projects.  So instead of num projects + 1 API calls, we would just have 1.

**Comment by tihuan** [comment]
From the Refinement, I believe the next step is to do a spike to figure out what combination of the solutions we want to try, and then create a ticket for that so we can estimate it. Is that what everyone understands too?

As for priority, judging from what @brianraymor and @ambrosejcarr said, it sounds like a P2?

Thank you!

**Comment by brianraymor** [comment]
When considering performance, please be aware that the portal will transition from a simple list of datasets to a list of collections with related metadata extracted from the datasets in the collection as outlined in [Consumers want to explore published collections](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/72).

**Comment by brianraymor** [comment]
Per conversation with @maniarathi - _We are migrating off of Lambda in mid-March which should solve the problem._

We'll create a new issue if that's not the case since this is related to the earlier dataset view.
